system-contracts/test/L1Messenger.spec.ts

```diff
@@ -0,0 +1,517 @@
+    stateDiffsSetupData = await setupStateDiffs();
+    logData = setupLogData();
+    logs = createLogs(l1MessengerAccount, l1Messenger, logData);
+    bytecodeData = await setupBytecodeData(l1Messenger.address);
+    await setResult("SystemContext", "txNumberInBlock", [], {
+      failure: false,
+      returnData: ethers.utils.defaultAbiCoder.encode(["uint16"], [1]),
+    });
+    emulator = new L1MessengerPubdataEmulator();
+  });
+
+  // this part is necessary to clean the state of L1Messenger contract
+  after(async () => {
+    emulator.addLog(logs[0]);
+    const fourthLog = ethers.utils.concat([
+      ethers.utils.hexlify([0]),
+      ethers.utils.hexlify([0]),
+      ethers.utils.hexZeroPad(ethers.utils.hexlify(1), 2),
+      ethers.utils.hexZeroPad(l1MessengerAccount.address, 20),
+      logData.key,
+      logData.value,
+    ]);
+    emulator.addLog(fourthLog);
+    emulator.addLog(logs[1]);
+    emulator.addMessage({ lengthBytes: logData.currentMessageLengthBytes, content: logData.message });
+    emulator.addBytecode(bytecodeData);
+    await l1Messenger
+      .connect(bootloaderAccount)
+      .publishPubdataAndClearState(emulator.buildTotalL2ToL1PubdataAndStateDiffs());
+    await network.provider.request({
+      method: "hardhat_stopImpersonatingAccount",
+      params: [TEST_L1_MESSENGER_SYSTEM_CONTRACT_ADDRESS],
+    });
+    await network.provider.request({
+      method: "hardhat_stopImpersonatingAccount",
+      params: [TEST_KNOWN_CODE_STORAGE_CONTRACT_ADDRESS],
+    });
+    await network.provider.request({
+      method: "hardhat_stopImpersonatingAccount",
+      params: [TEST_BOOTLOADER_FORMAL_ADDRESS],
+    });
+  });
+
+  describe("publishPubdataAndClearState", async () => {
+    it("publishPubdataAndClearState passes correctly", async () => {
+      await (
+        await l1Messenger.connect(l1MessengerAccount).sendL2ToL1Log(logData.isService, logData.key, logData.value)
+      ).wait();
+      emulator.addLog(logs[0]);
+      await (await l1Messenger.connect(l1MessengerAccount).sendToL1(logData.message)).wait();
+      emulator.addLog(logs[1]);
+      emulator.addMessage({ lengthBytes: logData.currentMessageLengthBytes, content: logData.message });
+      await (
+        await l1Messenger
+          .connect(knownCodeStorageAccount)
+          .requestBytecodeL1Publication(await ethers.utils.hexlify(utils.hashBytecode(bytecodeData.content)), {
+            gasLimit: 130000000,
+          })
+      ).wait();
+      emulator.addBytecode(bytecodeData);
+      emulator.setStateDiffsSetupData(stateDiffsSetupData);
+      await (
+        await l1Messenger
+          .connect(bootloaderAccount)
+          .publishPubdataAndClearState(emulator.buildTotalL2ToL1PubdataAndStateDiffs(), { gasLimit: 10000000 })
+      ).wait();
+    });
+
+    it("should revert Too many L2->L1 logs", async () => {
+      // set numberOfLogsBytes to 0x900 to trigger the revert (max value is 0x800)
+      emulator.numberOfLogs = 0x900;
```

Comment:
Do you think it would make sense to create a new `emulator` for each test? it is generally desirable to have tests are independent from one another, but LMK if you believe it introduces too many bloat

Reply:
I believe we should keep the global emulator, as it behaves similarly to the contract itself in the current tests.  We could be to recreate new instances of the emulator for each test,  which would involve additional code in creating valid data for the emulator in each test case, since we have to keep track with the real data of L1Messenger. That's why I'd suggest keeping current solution.

Reply:
@Uacias okay, global emulator is fine. I see the following pattern multiple times:

- Amend a field of the emulator 
- (Try and fail)
- Return the field to the previous value. 

The problem with how it is written right now is that if in the future we'd want to add a test, we'll have to amend all after the added one (as the "previous" value would change). 

I would suggest doing either of the following instead of this:
- Support overrides in the `buildTotalL2ToL1PubdataAndStateDiffs` (basically the ability to provide some of the fields). This approach is preferred IMHO. 
- If the first approach is hard, at least not rely on constant, but remember the previous from the state of `emulator`, not with a constant